packages/kolme/src/gossip/sync_manager.rs

```diff
@@ -0,0 +1,623 @@
+// ! Implementation of the sync manager state machine.
+//
+// Due to the complexity of the topic, there's a dedicated doc for this on the docs
+// site called "Sync manager". Recommendation: read that before reading
+// the code below.
+use std::{
+    collections::{hash_map, HashMap, HashSet},
+    fmt::Display,
+    time::Instant,
+};
+
+use gossip::{ReportBlockHeight, Trigger};
+use libp2p::PeerId;
+use smallvec::SmallVec;
+use utils::trigger::TriggerSubscriber;
+
+use crate::*;
+
+/// Status of state syncing.
+pub(super) struct SyncManager<App: KolmeApp> {
+    /// Blocks we're currently waiting on.
+    needed_blocks: BTreeMap<BlockHeight, WaitingBlock<App::Message>>,
+    /// Trigger [Gossip] to recheck sync requests.
+    trigger: Trigger,
+}
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug)]
+enum WaitingBlock<AppMessage> {
+    /// We haven't received the basic block info yet.
+    Needed(RequestStatus),
+    /// We have the raw block, but haven't processed it yet.
+    Received {
+        block: Arc<SignedBlock<AppMessage>>,
+        peer: Option<PeerId>,
+    },
+    /// We've received the block info, but now need to download the state layers.
+    Pending(PendingBlock<AppMessage>),
+}
+
+#[derive(Debug)]
+struct PendingBlock<AppMessage> {
+    block: Arc<SignedBlock<AppMessage>>,
+    /// Layers we're interested in but haven't received any info on yet.
+    needed_layers: BTreeMap<Sha256Hash, RequestStatus>,
```

Comment:
Do you think it would be better to use a data structure that preserves the insertion order ?  So that when Merkle layer's children are inserted, and then processed in the order in which they were inserted. Or it doesn't matter ?

Reply:
There might be some advantages, but I'm not sure what they'd be. If anything, I'd sort based on how deep in the map something is to try and get as many leaf layers added first, followed by their direct dependents, to reduce in memory storage. But I'm not sure it's worth the extra complexity we'd incur in managing the list. It may be worth reviewing when we have more real world data.